File: 剑指offer/剑指 Offer 53 - II. 0～n-1中缺失的数字.cpp
/*
 * @Description: 
 * @Version: 1.0
 * @Author: Vicro
 * @Date: 2020-12-03 22:29:07
 * @LastEditTime: 2020-12-03 22:43:42
 * @FilePath: \Leetcode\剑指 Offer 53 - II. 0～n-1中缺失的数字.cpp
 */

/*
RESULT: Accept
TIME: 32ms BEAT: 95.24%
MEMORY: 17.2MB BEAT: 16.24%
Description: 二分法
*/
class Solution {
public:
    int missingNumber(vector<int>& nums) {
        int l = 0, r = nums.size() - 1;
        
        while (l < r)
        {
            int mid = (r - l) / 2 + l;

            if (nums[mid] == mid)   l = mid + 1;
            else r = mid;
        }

        if (nums[l] == l) return ++l;
        else return l;
    }
};

File: 剑指offer/剑指 Offer 53 - II. 0～n-1中缺失的数字.py
'''
Description: 
Version: 1.0
Author: 冰凝水
Date: 2020-12-03 22:45:32
LastEditTime: 2020-12-03 22:48:19
FilePath: \Leetcode\剑指 Offer 53 - II. 0～n-1中缺失的数字.py
'''

"""
RESULT: Accept
TIME: 40ms BEAT: 88.02%
MEMORY: 14.5MB BEAT: 28.29%
Description: 二分查找
"""
class Solution:
    def missingNumber(self, nums: List[int]) -> int:
        l, r = 0, len(nums) - 1
        
        while l < r:
            mid = (r - l) // 2 + l;

            if nums[mid] == mid:
                l = mid + 1
            else:
                r = mid
        
        if nums[l] == l:
            return l + 1
        else:
            return l